Step1 - 인수 테스트 기반 리팩터링

mins99 opened the pull request:
안녕하세요 리뷰어님! 처음 뵙겠습니다. 김민성입니다 : )
step1 리팩터링 완료하여 PR 요청드립니다.

미션 요구사항의 순서에 맞춰 리팩터링을 진행해보았는데, 
테스트의 보호속에서 개발해보는 경험을 할 수 있었습니다 🥳
아무래도 LineService에 대한 리팩터링 과제이다 보니 line, station 패키지 위주로 개선했습니다.
추가로 더 개선하면 좋은 부분이 있다면 편하게 말씀해주세요. 개선하도록 하겠습니다

남은 미션들도 각 요구사항을 최대한 신경쓰면서 많이 배우고 가겠습니다 🙇🏻
미션 잘 부탁드립니다 🥰

mins99 committed:
docs(README): 1단계 - 인수 테스트 기반 리팩터링 요구 사항 정리

mins99 committed:
feat(Distance): 거리 원시값 포장

mins99 committed:
refactor(Section): 거리 객체(Distance) 적용

mins99 committed:
test(Distance): 거리 객체 테스트 추가

mins99 committed:
refactor(getStations): 역 리스트를 가져오는 로직을 Line으로 이동

mins99 committed:
feat(Section): 테스트용 생성자 추가

mins99 committed:
feat(Station): 테스트용 생성자 추가

mins99 committed:
feat(Sections): Section List 일급 컬렉션 추가

mins99 committed:
refactor(Line): 구간 조회 로직 개선

- 일급컬렉션(Sections) 추가
- 역 리스트를 가져오는 getStations을 Sections로 이동

mins99 committed:
test(SectionsTest): 구간 객체 테스트 추가

mins99 committed:
feat(Line): 서비스 로직 개선

- 읽기 전용 트랜잭션 구분
- 공통 메서드 분리
- 데이터를 못찾는 경우 NotFountException 으로 처리

mins99 committed:
feat(Station): 서비스 로직 개선

- 읽기 전용 트랜잭션 구분
- 공통 메서드 분리
- 데이터를 못찾는 경우 NotFountException 으로 처리

mins99 committed:
refactor(addLineStation): 구간 추가 로직 개선

- 서비스 로직 도메인 영역으로 변경(Sections, Section, Line)
- 도메인 테스트 추가

mins99 committed:
refactor(removeLineStation): 구간 삭제 로직 개선

- 서비스 로직 도메인 영역으로 변경(Sections, Line)
- 도메인 테스트 추가

mins99 committed:
docs(README): 1단계 - 인수 테스트 기반 리팩터링 구현 사항 체크

mins99 committed:
refactor(Sections): 역 조회 로직 수정

- 반복문 depth 1 줄이기

mins99 committed:
refactor(Line): 미사용 메서드 삭제

mins99 committed:
test(LineSectionAcceptanceTest): 인수 테스트 통합

mins99 committed:
docs(README): 1단계 - 인수 테스트 기반 리팩터링 구현 사항 체크

dhmin5693 commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -29,154 +24,55 @@ public LineService(LineRepository lineRepository, StationService stationService)
-
-        return downStation;
+    private Line findLineById(Long id) {
+        return lineRepository.findById(id).orElseThrow(NotFoundException::new);
```
`RuntimeException`을 적절하게 바꿔준건 매우 좋네요 👍 
다만 그냥 예외를 발생시키기보다는 메시지에 id 등을 적당히 첨가해주면 오류 원인을 찾을 때 좋을 것 같아요.

mins99 replied:
[커스텀 예외 효과적으로 사용하기](https://seovalue.github.io/2022/02/21/custom-exception/)
예외처리를 잘 활용하지 못하고 있었는데 좋은 공부가 되었습니다 😆

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -51,22 +63,33 @@ public Station getDownStation() {
-        }
-        this.downStation = station;
-        this.distance -= newDistance;
+    public boolean hasAllStations(Section section) {
```
equals로 정의하면 어떨까요? 사실상 그 역할을 하고 있는 것으로 보입니다.

mins99 replied:
`hasAllStations`, `hasAnyStations` 의 경우 어느 정도로 리팩터링을 하는게 최선일지 고민이 되었던 부분인데요..
네 번의 비교가 필요한 메서드인데, 하나의 메서드에서 처리해도 괜찮을까요? 🤔
우선은 각 메서드에서 처리하도록 수정해보았습니다!

dhmin5693 replied:
의미를 잘못 전달드린 것 같아요.

`hasAllStation`를 통해 어떤 결과를 얻어가려고 하셨나요?
코드로 추측을 해보자면 `상행/하행역이 완전히 같은지`를 확인하려고 하셨고
이미 같은 상, 하행을 가진 `Section`이 있다면 추가되지 않도록 처리되어 있습니다.

그렇기 때문에 영속성이 보장되는 상태에서 `Section` 도메인은 상/하행만 같아도 실질적으로 동등하다고 생각할 수 있습니다.

mins99 replied:
아 equals의 의미가 Section에 대한 비교를 의미하신거군요 2단계에서 해당 부분에 대해 개선해보겠습니다!

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -51,22 +63,33 @@ public Station getDownStation() {
-    public void updateUpStation(Station station, int newDistance) {
-        if (this.distance <= newDistance) {
-            throw new RuntimeException("역과 역 사이의 거리보다 좁은 거리를 입력해주세요");
+    public void changeStationInfo(Section section) {
```
개인적인 의견이지만 `info`, `data`라는 postfix 는 대부분의 경우에서 있어도 그만, 없어도 그만이더라구요.
그래서 저는 가급적이면 사용하지 않으려고 하는 용어입니다.

`changeStation`, `changeStationInfo` 두 개를 놓고 보면 어떤가요?

mins99 replied:
`info`만 봤을때 어떤 데이터를 수정하는지도 알수 없으니 굳이 붙일 필요가 없었던것 같습니다 😅

dhmin5693 commented on src/main/java/nextstep/subway/line/dto/LineResponse.java:
```diff
@@ -26,8 +27,8 @@ public LineResponse(Long id, String name, String color, List<StationResponse> st
-    public static LineResponse of(Line line, List<StationResponse> stations) {
-        return new LineResponse(line.getId(), line.getName(), line.getColor(), stations, line.getCreatedDate(), line.getModifiedDate());
+    public static LineResponse of(Line line) {
+        return new LineResponse(line.getId(), line.getName(), line.getColor(), createStationResponse(line), line.getCreatedDate(), line.getModifiedDate());
```
생성자가 여기서만 쓰이기 때문에 private으로 바꾸어도 괜찮겠군요.

dhmin5693 commented on src/main/java/nextstep/subway/line/ui/LineController.java:
```diff
@@ -64,4 +64,9 @@ public ResponseEntity removeLineStation(@PathVariable Long lineId, @RequestParam
         return ResponseEntity.badRequest().build();
     }
+
+    @ExceptionHandler(IllegalArgumentException.class)
```
NotFoundException.class 도 추가해줘야 하지 않을까요?
`Line`이나 `Station`을 못찾으면 500 오류로 반환하게 됩니다.

mins99 replied:
커스텀 클래스를 생성하고 깜빡했네요😱 수정하고 테스트까지 완료했습니다 : )
수정하다보니 바로 위 메서드 `IllegalArgumentException.class`와 관련해서 고민(?)이 새로 생겼는데요
리뷰어님은 둘 중 어떤 메서드를 선호하시는지 궁금합니다!
```
ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
ResponseEntity.badRequest().body(e.getMessage());
```

dhmin5693 replied:
주신 두 가지 선택지에선 아래가 더 보기 좋다고 생각해요.
저는 보통 `@ResponseStatus` 어노테이션을 사용합니다.

mins99 replied:
`@ResponseStatus`에 대해서도 2단계 미션에서 학습해보겠습니다 감사합니다 🥰

dhmin5693 commented on src/main/java/nextstep/subway/station/domain/Station.java:
```diff
@@ -13,10 +13,15 @@ public class Station extends BaseEntity {
+        this(null, name);
+    }
+
+    public Station(Long id, String name) {
```
entity 인스턴스가 영속 상태라면 이 인스턴스는 어디로부터 왔을까요?
보통은 save 등을 통해 저장을 하거나 find로 찾아왔을 때입니다.
JPA를 사용할 땐 영속 상태인지, 준영속, 비영속 상태인지를 항상 염두에 두어야 예기치 못한 오류를 방어할 수 있어요.

영속 상태를 확실하게 보장하기 위해선 생성자에서 id를 제거해주는게 좋습니다.
만약 테스트를 위해 만드신 거라면 Fixture를 따로 생성해주세요.
id는 `ReflectionTestUtils.setField();` 으로 설정할 수 있습니다.

mins99 replied:
도메인 테스트시에는 따로 영속상태로 만들지 않아서 만들었던 생성자인데 삭제후 테스트해보니 굳이 필요없는 파라미터였네요
오히려 영속성에 영향을 줄 수 있는 로직이 될 거라고는 생각 못했습니다 😲
알려주신 `ReflectionTestUtils`도 새로 학습했습니다 감사합니다 🙇🏻

dhmin5693 commented on src/test/java/nextstep/subway/line/acceptance/LineSectionAcceptanceTest.java:
```diff
@@ -113,6 +122,33 @@ void removeLineSection2() {
+     *     When 지하철 노선에 등록된 역 목록 조회 요청
+     *     Then 삭제한 지하철 구간이 반영된 역 목록이 조회됨
+     */
+    @DisplayName("지하철 구간을 관리")
```
`지하철 구간을 관리` 라는 시나리오로는 이 테스트가 무엇을 위해 작성되었는지 파악하기 어렵습니다.
`DisplayName`에는 원인과 결과가 명확한 문장으로 적어주세요.

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -0,0 +1,62 @@
+        return distance;
+    }
+
+    public Distance addDistance(Distance targetDistance) {
```
`add` 만 적어도 충분히 의미를 전달할 수 있다고 생각해요.
`subtract`도 마찬가지입니다.

dhmin5693 commented on src/test/java/nextstep/subway/line/domain/SectionTest.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    @Test
+    @DisplayName("해당 구간이 다른 구간과 완전히 동일하지 않다")
```
`동일하다`를 어떻게 정의했는지를 명시하면 더 보기 편하지 않을까요?

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,109 @@
+            stations.add(downStation);
+        }
+
+        return stations;
```
반환한 `stations`를 외부에서 가공하면(stations.add 등) 이 메소드의 의미가 퇴색될거에요.
읽기 전용으로 내보내는건 어떤가요?

mins99 replied:
데이터를 가져오는 것만 신경썼네요😅 
불변🔥을 지킬 수 있도록 수정했습니다!

dhmin5693 reviewed (changes requested):
안녕하세요, 민성님.
ATDD&TDD 미션을 지원해드릴 민동휘라고 합니다.
반갑습니다 👋

로직 자체는 리팩토링을 굉장히 잘 해주셨습니다 👍 
조금 더 깔끔한 코드를 위해 작은 의견을 몇 개 남겨두었으니 확인 부탁드립니다.
확인 후 재요청 부탁드릴게요 🙏

mins99 committed:
refactor(NotFoundException): 커스텀 예외 효과적으로 사용하기

mins99 committed:
refactor(Section): hasUpStation, hasDownStation 메서드 삭제

mins99 committed:
refactor(Section): changeStationInfo 메서드명 수정

- 메서드에 info, data 와 같은 단어를 지양하자

mins99 committed:
refactor(constructor): 생성자 메서드 접근 제어자 변경

- of 메서드로 생성하므로 기본 생성자는 private 변경

mins99 committed:
refactor(LineController): 커스텀 예외 효과적으로 사용하기2

mins99 committed:
refactor(constructor): 테스트용 생성자 삭제

- 영속 상태를 확실하게 보장하기 위해선 생성자에서 id를 제거해야 함
- 테스트에 id가 필요하다면 ReflectionTestUtils.setField(); 사용

mins99 committed:
refactor(LineSectionAcceptanceTest): 메서드 설명 추가

- DisplayName에는 원인과 결과가 명확한 문장으로 작성

mins99 committed:
refactor(Distance): 메서드명 변경

mins99 committed:
refactor(SectionTest): DisplayName의 내용을 좀 더 상세하게 작성

mins99 committed:
refactor(Sections): 조회 메서드 읽기 전용으로 반환하도록 변경

- 불변🔥

mins99 commented:
안녕하세요 리뷰어님!
약간의 슬럼프와😥 좋지않은 컨디션으로 피드백 반영이 늦었습니다
(허약한 체력을 기르려고 운동을 하는데 운동을 하니 몸이 아프네요 😅)

피드백 주신 내용 중 일부는 사실 3주차 ATDD 미션하면서 동일하게 작성했던 로직도 있었는데 
찜찜했던 부분들이 많이 개선되어 기쁜 마음입니다

확인 부탁드릴게요 🥳

dhmin5693 reviewed (approved):
피드백 잘 반영해주셨네요 :)

아무래도 우테캠 프로 교육이 2개월간 상당히 빡빡한 스케줄을 요구하다 보니 체력적으로 어려우실 수 있어요.
건강이 우선입니다.
충분한 휴식을 취해야 코드도 더 잘 나오더라구요 😄

1단계는 이걸로 종료해도 될 것 같아요.
몇 가지 코멘트를 남겨놨습니다.
unresolve review를 천천히 확인해주시고, 다음 미션 진행해주세요.
